cabal-install/Distribution/Client/ProjectBuilding.hs

```diff
@@ -1133,10 +1133,11 @@ buildInplaceUnpackedPackage verbosity
       | elabBuildHaddocks pkg = action
       | otherwise            = return ()
 
-    whenReRegister  action = case buildStatus of
-      BuildStatusConfigure          _ -> action
-      BuildStatusBuild Nothing      _ -> action
-      BuildStatusBuild (Just mipkg) _ -> return mipkg
+    whenReRegister  action
+      = case buildStatus of
+          BuildStatusBuild (Just mipkg) _ -> return mipkg
+          _ | null (elabBuildTargets pkg) -> return Nothing -- ick
```

Comment:
Yeah, this line highlights the difficulty. Consider the case of

     BuildStatusBuild Nothing _
      | null (elabBuildTargets pkg) -> ...

So in this case there's no previous registration and normally we'd try and register, but because we're not doing a full build for this component then actually we still don't want to register. But there's no way to represent this because the Nothing result here means that the thing is an executable with no registration, not that we're skipping the registration phase.

Now I had previously expected that we'd solve this problem by adjusting the planning phase to say that registration is not required, i.e. `elabRequiresRegistration = False`, but now I see that's going to have the same issue as above, and would cause problems for keeping track of whether we did register or not if we do `cabal repl; cabal build`.

I think either way the solution is going to have to involve extending the BuildResult's buildResultLibInfo so that it can properly represent the situation of not having registered yet, distinct from registration not needed for exes. 

We'll have the same issue once we support building only specific modules.

I've also been working on build status, and this feels related, so I'll have another think about it in that context.